mkallTests/MKReporterTests.mm

```diff
@@ -0,0 +1,231 @@
+})";
+
+static const char *thirdMeasurement = R"({
+  "data_format_version": "0.2.0",
+  "input": "kernel.org",
+  "measurement_start_time": "2016-06-04 17:53:13",
+  "probe_asn": "AS0",
+  "probe_cc": "ZZ",
+  "probe_ip": "127.0.0.1",
+  "software_name": "measurement_kit",
+  "software_version": "0.2.0-alpha.1",
+  "test_keys": {"failure": null,"received": [],"sent": []},
+  "test_name": "gummy",
+  "test_runtime": 0.253494024276733,
+  "test_start_time": "2016-06-04 17:53:13",
+  "test_version": "0.0.1"
+})";
+
+// Utility function to create a task with specific configuration
+static MKReporterTask *makeTask(
+    NSString *softwareName, NSString *softwareVersion) {
+  return [[MKReporterTask alloc] initWithSoftwareName:softwareName
+                                      softwareVersion:softwareVersion];
+}
+
+// Utility function to create a measurement from a C string
+static NSString *makeMeasurement(const char *m) {
+  return [NSString stringWithUTF8String:m];
+}
+
+@interface MKReporterTests : XCTestCase
+@end // @interface MKReporterTests
+
+@implementation MKReporterTests
+
+// Utility function to submit a measurement with an existing task.
+- (void)submissionWithTask:(MKReporterTask *)task
+               measurement:(NSString *)measurement
+            expectedResult:(BOOL)expectedResult
+             expectedStats:(NSDictionary *)expectedStats {
+  NSMutableDictionary *stats = [[NSMutableDictionary alloc] init];
+  MKReporterResults *results = [task submitWithMeasurement:measurement
+                                             uploadTimeout:0
+                                                     stats:stats];
+  NSLog(@"logs: %@", [results logs]);
+  NSLog(@"=== BEGIN VARIABLES ===");
+  NSLog(@"good: %d", [results good]);
+  NSLog(@"updatedSerializedMeasurement: %@", [results updatedSerializedMeasurement]);
+  NSLog(@"updatedReportID: %@", [results updatedReportID]);
+  NSLog(@"=== END VARIABLES ===");
+  XCTAssert([[results logs] length] > 0);
+  XCTAssert([results good] == expectedResult);
+  if ([results good]) {
+    NSString *rid = [results updatedReportID];
+    XCTAssert([rid length] > 0);
+    XCTAssert([[results updatedSerializedMeasurement] containsString:rid]);
+  }
+  XCTAssert([stats isEqualToDictionary:expectedStats]);
+}
+
+// Utility function to submit just a single measurement.
+- (void)simpleSubmission:(NSString *)measurement
+            softwareName:(NSString *)softwareName
+         softwareVersion:(NSString *)softwareVersion
+          expectedResult:(BOOL)expectedResult
+           expectedStats:(NSDictionary *)expectedStats {
+  return [self submissionWithTask:makeTask(softwareName, softwareVersion)
+                      measurement:measurement
+                   expectedResult:expectedResult
+                    expectedStats:expectedStats];
+}
+
+// Ensure that we can submit a single measurement.
+- (void)testResubmissionGood {
+  NSDictionary *stats = @{
+    @"bouncer_error": @0,
```

Comment:
shouldn't @ be only in front of string? I you want an int should be only
`@"bouncer_error": 0`, or am I missing something?

Reply:
The `@` in front of a number creates a `NSNumber`. I need to insert `NSNumbers` there mainly because that is what I do in the code that measures what happens.

Reply:
didn't know that, I always used `[NSNumber numberWithDouble:1000]`